refactor(inspector): refactor entity properties + window registry + docking registry

iMeaNz opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new material inspector and enhanced property panels for better scene and entity interaction.
  
- **Refactor**
  - Streamlined scene management and window registration, providing a more consistent and intuitive user interface.
  - Improved window docking and layout behavior for easier navigation.

- **Chores**
  - Removed obsolete logging functionality and reorganized internal code for enhanced maintainability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

iMeaNz commented:
# Inspector and Entity properties refactor
Before we were using static classes to render the entity properties. But that caused problem when the properties were not pure and were supposed to be able to modify the global state.
This PR introduces changes on the way we handle entity properties and the inspector window

# Entity properties
Every entity properties now inherits from an abstract class `AEntityProperty` that inherits from an interface `IEntityProperty`
```cpp
class IEntityProperty {
    public:
        virtual ~IEntityProperty() = default;
        virtual bool show(ecs::Entity entity) = 0;
};

class AEntityProperty : public IEntityProperty {
    public:
        AEntityProperty(InspectorWindow &inspector) : m_inspector(inspector) {};
    protected:
        InspectorWindow &m_inspector;
};
```
The `AEntityProperty` constructor takes as argument the main **Inspector window** as a reference, this allow non-pure properties to be able to modify the global state. More on this in the next section

# Inspector Window
The Inspector has been modified to now hold shared pointers to `IEntityProperty`
```cpp
std::unordered_map<std::type_index, std::shared_ptr<IEntityProperty>> m_entityProperties;
```
For now the **Inspector Window** still holds the **Material Inspector** but this will change later in the PR.
This is how we setup evertyhing now : 
```cpp
InspectorWindow::InspectorWindow()
{
   	m_materialInspector = std::make_shared<MaterialInspector>(*this);

	m_entityProperties[typeid(components::TransformComponent)] = std::make_shared<TransformProperty>(*this);
	m_entityProperties[typeid(components::RenderComponent)] = std::make_shared<RenderProperty>(*this);
	m_entityProperties[typeid(components::AmbientLightComponent)] = std::make_shared<AmbientLightProperty>(*this);
	m_entityProperties[typeid(components::DirectionalLightComponent)] = std::make_shared<DirectionalLightProperty>(*this);
	m_entityProperties[typeid(components::PointLightComponent)] = std::make_shared<PointLightProperty>(*this);
	m_entityProperties[typeid(components::SpotLightComponent)] = std::make_shared<SpotLightProperty>(*this);
	m_entityProperties[typeid(components::CameraComponent)] = std::make_shared<CameraProperty>(*this);
	m_entityProperties[typeid(components::PerspectiveCameraController)] = std::make_shared<CameraController>(*this);
}
```
Furthermore, we now also have a bunch of utils functions such as :
```cpp
/**
* @brief Sets the visibility flag for the sub-inspector identified by type T.
*
* @tparam T The sub-inspector type.
* @param visible True if the sub-inspector should be visible, false otherwise.
*/
template<typename T>
void setSubInspectorVisibility(bool visible)
{
    m_subInspectorVisibility[std::type_index(typeid(T))] = visible;
}

/**
* @brief Sets the associated data for the sub-inspector identified by type T.
*
* @tparam T The sub-inspector type.
* @param data Pointer to a components::Material instance, or nullptr.
*/
template<typename T>
void setSubInspectorData(components::Material *data)
{
    m_subInspectorData[std::type_index(typeid(T))] = data;
}

/**
* @brief Gets the visibility flag for the sub-inspector identified by type T.
*
* @tparam T The sub-inspector type.
* @return bool True if the sub-inspector is marked as visible; false if not found.
*/
template<typename T>
bool getSubInspectorVisibility() const
{
    auto it = m_subInspectorVisibility.find(std::type_index(typeid(T)));
    return (it != m_subInspectorVisibility.end()) ? it->second : false;
}

/**
* @brief Gets a modifiable reference to the visibility flag for the sub-inspector of type T.
*
* This function returns a reference to the boolean stored in m_subInspectorVisibility
* associated with the sub-inspector type T. If the entry does not exist, it is default-initialized
* (to false) and inserted into the map.
*
* @tparam T The type representing the sub-inspector.
* @return bool& A modifiable reference to the visibility flag.
*/
template<typename T>
bool& getSubInspectorVisibility()
{
    return m_subInspectorVisibility[std::type_index(typeid(T))];
}

/**
* @brief Gets the associated data for the sub-inspector identified by type T.
*
* The returned variant will either contain a pointer to a components::Material or std::monostate
* if no data was set.
*
* @tparam T The sub-inspector type.
* @return std::variant<std::monostate, components::Material*> The stored data.
*/
template<typename T>
std::variant<std::monostate, components::Material*> getSubInspectorData() const
{
    auto it = m_subInspectorData.find(std::type_index(typeid(T)));
    return (it != m_subInspectorData.end()) ? it->second : std::variant<std::monostate, components::Material*>{std::monostate{}};
}
```
Those allows the entity properties to handle global state while also allowing other document windows to access information about the modification requested by the user.

iMeaNz commented:
# Window registry
We now have a window registry class.
This class is responsible for handling the different windows and providing easy access between them.
Each **Document Window** holds a reference to it and can access the data of any type of window.
```cpp
class WindowRegistry {
	public:
		template<typename T>
		void registerWindow(std::shared_ptr<T> window)
		{
			m_windows[typeid(T)] = window;
		}

		template<typename T>
		std::shared_ptr<T> getWindow()
		{
			if (m_windows.find(typeid(T)) == m_windows.end())
			{
				throw std::runtime_error("Window not registered");
			}
			return std::static_pointer_cast<T>(m_windows[typeid(T)]);
		}

		void setup();
		void shutdown();
		void update();
		void render();

	private:
		std::unordered_map<std::type_index, std::shared_ptr<IDocumentWindow>> m_windows;
};
```
Here is the default setup, we can register windows per-type and access them per-type.  Internally, the document windows directly call this class.
But externally, the editor abstract this setup with those two functions  :
```cpp
template<typename T>
void registerWindow()
{
   	auto window = std::make_shared<T>(m_windowRegistry);
   	m_windowRegistry.registerWindow<T>(window);
}

template<typename T>
std::shared_ptr<T> getWindow()
{
   	return m_windowRegistry.getWindow<T>();
}
```

Here is an example of how to use it:
```cpp
editor.registerWindow<nexo::editor::SceneViewManager>();
editor.registerWindow<nexo::editor::SceneTreeWindow>();
editor.registerWindow<nexo::editor::InspectorWindow>();
editor.registerWindow<nexo::editor::ConsoleWindow>();

editor.getWindow<nexo::editor::SceneViewManager>()->addNewScene("Default scene", true);
```

And internally:
```cpp
auto viewManager = m_windowRegistry.getWindow<SceneViewManager>();
viewManager->setSelectedScene(obj.data.sceneProperties.sceneId);
```

iMeaNz commented:
# Docking Registry
Added a new docking registry class that holds the dock ids of the different windows based on their names.
```cpp
class DockingRegistry {
	public:
		void setDockId(const std::string& name, ImGuiID id);
		ImGuiID getDockId(const std::string& name) const;

	private:
		std::unordered_map<std::string, ImGuiID> dockIds;
};
```
The `WindowRegistry` class holds a reference to this class so that every `ADocumentWindow` can access.
This is used when first rendering any window to make sure the current dock id is the one saved in the docking registry.
```cpp
void firstDockSetup(const std::string &windowName)
{
    if (ImGuiWindow* currentWindow = ImGui::GetCurrentWindow(); m_firstOpened && currentWindow)
    {
		const bool isDocked = currentWindow->DockIsActive;
		const ImGuiID currentDockID = currentWindow->DockId;

		if (!isDocked || currentDockID != m_windowRegistry.getDockId(windowName))
			currentWindow->DockId = m_windowRegistry.getDockId(windowName);
		m_firstOpened = false;
    }
}
```
The `DockingRegistry` is filled once at startup, either via the config file, or via default value if the config file does not exists.
```cpp
ImGui::DockBuilderDockWindow("Default scene", mainSceneTop);
ImGui::DockBuilderDockWindow("Console", consoleNode);
ImGui::DockBuilderDockWindow("Scene Tree", sceneTreeNode);
ImGui::DockBuilderDockWindow("Inspector", inspectorNode);
ImGui::DockBuilderDockWindow("Material Inspector", materialInspectorNode);

m_windowRegistry.setDockId("Default scene", mainSceneTop);
m_windowRegistry.setDockId("Console", consoleNode);
m_windowRegistry.setDockId("Scene Tree", sceneTreeNode);
m_windowRegistry.setDockId("Inspector", inspectorNode);
m_windowRegistry.setDockId("Material Inspector", materialInspectorNode);
```
```cpp
m_windowRegistry.setDockId("Default scene", findWindowDockIDFromConfig("Default scene"));
m_windowRegistry.setDockId("Console", findWindowDockIDFromConfig("Console"));
m_windowRegistry.setDockId("Scene Tree", findWindowDockIDFromConfig("Scene Tree"));
m_windowRegistry.setDockId("Inspector", findWindowDockIDFromConfig("Inspector"));
m_windowRegistry.setDockId("Material Inspector", findWindowDockIDFromConfig("Material Inspector"));
```
The dock id in the config is retrieved thanks to a small functions that parses the config file based on a string name (maybe in the futur make a function that retrieves every window stored in the config file)
```cpp
ImGuiID findWindowDockIDFromConfig(const std::string& windowName)
{
    std::string configPath = Path::resolvePathRelativeToExe(
        "../config/default-layout.ini").string();
    std::ifstream configFile(configPath);
    if (!configFile.is_open()) {
        std::cout << "Could not open config file: " << configPath << std::endl;
        return 0;
    }

    std::string line;
    bool inWindowSection = false;
    ImGuiID dockId = 0;

    std::string windowHeader = "[Window][" + windowName + "]";

    while (std::getline(configFile, line)) {
        // Found the window section we're looking for
        if (line == windowHeader) {
            inWindowSection = true;
            continue;
        }

        if (inWindowSection) {
            // If we hit a new section, stop searching
            if (!line.empty() && line[0] == '[')
                break;

            std::regex dockIdRegex("DockId=(0x[0-9a-fA-F]+)");

            if (std::smatch match; std::regex_search(line, match, dockIdRegex) && match.size() > 1) {
                std::string hexDockId = match[1];
                std::stringstream ss;
                ss << std::hex << hexDockId;
                ss >> dockId;
                break;
            }
        }
    }

    configFile.close();
    return dockId;
}
```

iMeaNz commented:
# Refactor Material Inspector
The **Material Inspector** now derives from the `ADocumentWindow` like any other window.
Everything works like before thanks to the new implementation of the `WindowRegistry` and `DockingRegistry` classes
It works like so :
```cpp
void MaterialInspector::show()
{
    auto const &selector = Selector::get();
    const int selectedEntity = selector.getSelectedEntity();
    auto inspectorWindow = m_windowRegistry.getWindow<InspectorWindow>();
	if (inspectorWindow->getSubInspectorVisibility<MaterialInspector>())
    {
        ImGuiWindowFlags window_flags = ImGuiWindowFlags_NoCollapse;
        if (m_firstOpened)
           	window_flags |= ImGuiWindowFlags_NoBringToFrontOnFocus;

        if (ImGui::Begin("Material Inspector", &inspectorWindow->getSubInspectorVisibility<MaterialInspector>(), window_flags))
        {
           	if (m_firstOpened)
          		firstDockSetup("Material Inspector");
            renderMaterialInspector(selectedEntity);
        }
        ImGui::End();
    }
}
```

## Notes
The window flags setup is here to avoid the window from poping instantly, the first frame it is rendering behind every other window, and then it is correctly rendered the second time

Thyodas commented:
@coderabbitai review

iMeaNz commented on editor/src/exceptions/Exceptions.hpp:
```diff
@@ -27,6 +29,12 @@ namespace nexo::editor {
+    	public:
+     		explicit WindowNotRegistered(std::type_index windowTypeIndex, const std::source_location loc = std::source_location::current())
+                : Exception(std::format("Window not registered: {}", windowTypeIndex.name()), loc) {}
+    };
```
Done

iMeaNz commented on editor/src/DocumentWindows/MainScene.cpp:
```diff
@@ -329,14 +331,14 @@
 	                    selector.setSelectionType(SelectionType::ENTITY);
                     }
-                    viewManager.setSelectedScene(m_sceneId);
+                    viewManager->setSelectedScene(m_sceneId);
```
Not necessary

iMeaNz commented on editor/src/DocumentWindows/MainScene.cpp:
```diff
@@ -329,14 +331,14 @@
                 {
                 	std::cout << "Clicked on entity with ID: " << data << std::endl;
-                    auto &viewManager = SceneViewManager::get();
+                    auto viewManager = m_windowRegistry.getWindow<SceneViewManager>();
```
Removed the print

iMeaNz commented on editor/src/DocumentWindows/MainScene.cpp:
```diff
@@ -33,11 +34,12 @@
+                                                                           m_defaultScene(defaultScene),
+                                                                           m_windowRegistry(windowRegistry)
+    {
+    	windowId = nextWindowId++;
```
Done

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.hpp:
```diff
@@ -0,0 +1,37 @@
+
+			void renderMaterialInspector(int selectedEntity);
+
+
+			std::shared_ptr<renderer::Framebuffer> m_framebuffer = nullptr;
+			int m_ecsEntity = -1;
```
Added a check

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.cpp:
```diff
@@ -0,0 +1,102 @@
+		framebufferSpecs.height = static_cast<unsigned int>(64);
+		m_framebuffer = renderer::Framebuffer::create(framebufferSpecs);
+		m_framebuffer->setClearColor({0.05f, 0.05f, 0.05f, 0.0f});
+	}
```
Not necessary for now

iMeaNz commented on editor/src/DocumentWindows/ConsoleWindow.cpp:
```diff
@@ -32,6 +32,20 @@
+            case LogLevel::DEV: return loguru::Verbosity_2;
+        }
+        return loguru::Verbosity_INVALID;
+    }
```
Done

iMeaNz commented on editor/src/Editor.hpp:
```diff
@@ -55,32 +31,36 @@
+            {
+            	auto window = std::make_shared<T>(m_windowRegistry);
+             	m_windowRegistry.registerWindow<T>(window);
+            }
```
Done

iMeaNz commented on editor/src/Editor.hpp:
```diff
@@ -55,32 +31,36 @@
             void setupStyle() const;
             void setupFonts(float scaleFactorX, float scaleFactorY) const;
-            void buildDockspace() const;
+            void buildDockspace();
```
Not necessary

iMeaNz commented on editor/src/Editor.cpp:
```diff
@@ -228,30 +174,11 @@ namespace nexo::editor {
-        return m_logs;
+		setupEngine();
+		setupStyle();
+		m_windowRegistry.setup();
```
Not necessary for now since we use exceptions

iMeaNz commented on editor/src/Editor.cpp:
```diff
@@ -337,20 +253,28 @@
         }
-
-        if (g_materialInspectorDockID == 0)
+        else if (!dockingRegistryFilled)
```
Could be implemented later

iMeaNz commented on editor/src/Editor.cpp:
```diff
@@ -265,31 +192,20 @@
     {
         const ImGuiViewport* viewport = ImGui::GetMainViewport();
         const ImGuiID dockspaceID = viewport->ID;
+        static bool dockingRegistryFilled = false;
```
Not necessary

iMeaNz commented on editor/src/Editor.cpp:
```diff
@@ -370,13 +294,7 @@
-            if (window->isOpened())
-                window->show();
-        }
+        m_windowRegistry.render();
```
Should be fine

iMeaNz commented on editor/src/Editor.cpp:
```diff
@@ -415,14 +333,9 @@
-        {
-            window->update();
-        }
+    	m_windowRegistry.update();
```
Should be fine

iMeaNz commented on editor/src/Editor.hpp:
```diff
@@ -55,32 +31,36 @@ namespace nexo::editor {
-            std::unordered_map<std::string, std::shared_ptr<IDocumentWindow>, TransparentStringHash, TransparentStringEqual> m_windows;
-
-            std::vector<LogMessage> m_logs;
+            WindowRegistry m_windowRegistry;
```
Should be fine

iMeaNz commented on editor/src/DocumentWindows/EntityProperties/RenderProperty.cpp:
```diff
@@ -214,17 +199,12 @@ namespace nexo::editor {
+         	m_popupManager.closePopup();
         }
 
         return true;
     }
```
Not necessary

iMeaNz commented on editor/src/DocumentWindows/ConsoleWindow.cpp:
```diff
@@ -54,19 +68,46 @@ namespace nexo::editor {
     void ConsoleWindow::setup()
-    {}
+    {
+    }
```
Done

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.cpp:
```diff
@@ -0,0 +1,102 @@
+	void MaterialInspector::shutdown()
+	{
+
+	}
```
Done, but the framebuffer reset is not necessary

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.cpp:
```diff
@@ -0,0 +1,102 @@
+	void MaterialInspector::update()
+	{
+
+	}
```
Done

iMeaNz commented:
@coderabbitai review

iMeaNz commented:
@coderabbitai generate docstrings

iMeaNz commented on editor/src/DockingRegistry.hpp:
```diff
@@ -0,0 +1,32 @@
+
+namespace nexo::editor {
+
+	class DockingRegistry {
+		public:
+			void setDockId(const std::string& name, ImGuiID id);
+			ImGuiID getDockId(const std::string& name) const;
+
+		private:
+			std::unordered_map<std::string, ImGuiID, TransparentStringHash, std::equal_to<>> dockIds;
+	};
```
Done

iMeaNz commented on editor/src/DocumentWindows/InspectorWindow.cpp:
```diff
@@ -39,68 +41,18 @@ extern ImGuiID g_materialInspectorDockID;
     void InspectorWindow::setup()
     {
-    	m_materialInspector->setup();
+		m_entityProperties[typeid(components::TransformComponent)] = std::make_shared<TransformProperty>(*this);
+		m_entityProperties[typeid(components::RenderComponent)] = std::make_shared<RenderProperty>(*this);
+		m_entityProperties[typeid(components::AmbientLightComponent)] = std::make_shared<AmbientLightProperty>(*this);
+		m_entityProperties[typeid(components::DirectionalLightComponent)] = std::make_shared<DirectionalLightProperty>(*this);
+		m_entityProperties[typeid(components::PointLightComponent)] = std::make_shared<PointLightProperty>(*this);
+		m_entityProperties[typeid(components::SpotLightComponent)] = std::make_shared<SpotLightProperty>(*this);
+		m_entityProperties[typeid(components::CameraComponent)] = std::make_shared<CameraProperty>(*this);
+		m_entityProperties[typeid(components::PerspectiveCameraController)] = std::make_shared<CameraController>(*this);
```
Not necessary for now

iMeaNz commented on editor/src/DocumentWindows/InspectorWindow.cpp:
```diff
@@ -196,9 +125,9 @@
+            if (m_entityProperties.contains(type))
             {
-                m_componentShowFunctions[type](entity);
+                m_entityProperties[type]->show(entity);
```
Should not happen

iMeaNz commented on editor/src/DocumentWindows/SceneViewManager.hpp:
```diff
@@ -31,17 +32,12 @@ namespace nexo::editor {
             void show() override;
             void update() override;
-            void addNewScene(const std::string &sceneName, const std::shared_ptr<MainScene>& scene = nullptr);
+            void addNewScene(const std::string &sceneName, bool defaultScene = false);
```
It is fine

iMeaNz commented on editor/src/DocumentWindows/SceneTreeWindow.cpp:
```diff
@@ -358,7 +358,7 @@ namespace nexo::editor {
 
         // Retrieves the scenes that are displayed on the GUI
-        auto scenes = SceneViewManager::get().getOpenScenes();
+        const auto &scenes = m_windowRegistry.getWindow<SceneViewManager>()->getOpenScenes();
```
Not necessary for now

iMeaNz commented on editor/src/DocumentWindows/SceneTreeWindow.cpp:
```diff
@@ -76,21 +74,21 @@
     void SceneTreeWindow::sceneSelected(const SceneObject &obj) const
     {
-        auto &viewManager = SceneViewManager::get();
+        auto viewManager = m_windowRegistry.getWindow<SceneViewManager>();
```
Should not happen

iMeaNz commented on editor/src/WindowRegistry.cpp:
```diff
@@ -0,0 +1,61 @@
+        }
+	}
+
+	void WindowRegistry::render() const
+	{
+		for (const auto &[_, window]: m_windows)
+        {
+            if (window->isOpened())
+                window->show();
+        }
+	}
```
It is called

iMeaNz commented on editor/src/WindowRegistry.cpp:
```diff
@@ -0,0 +1,61 @@
+        }
+	}
+
+	void WindowRegistry::setDockId(const std::string& name, ImGuiID id)
+	{
+		m_dockingRegistry.setDockId(name, id);
+	}
```
If name is empty it does not matter
And if it already exists, this should not be a problem

iMeaNz commented on editor/src/WindowRegistry.cpp:
```diff
@@ -0,0 +1,61 @@
+
+namespace nexo::editor {
+
+	void WindowRegistry::setup() const
+	{
+		for (const auto &[_, window]: m_windows)
+        {
+            window->setup();
+        }
+	}
```
Not necessary, we use exceptions

iMeaNz commented on editor/src/DocumentWindows/SceneViewManager.cpp:
```diff
@@ -28,10 +28,7 @@ namespace nexo::editor {
-        {
-            scene->setup();
-        }
+    	// No need to setup anything since the scene's setup are done when adding it
```
Not necessary

iMeaNz commented on editor/src/DocumentWindows/SceneViewManager.cpp:
```diff
@@ -58,23 +55,17 @@
-        {
-            newScene = std::make_shared<MainScene>(sceneName);
-            newScene->setup();
-        }
+        auto newScene = std::make_shared<MainScene>(m_windowRegistry, sceneName, defaultScene);
+        newScene->setup();
         const std::string targetWindow = !m_scenes.empty() ? m_scenes.begin()->second->getName() : "";
         if (!targetWindow.empty())
         {
             if (const ImGuiDockNode *dockNode = getDockNodeForWindow(targetWindow.c_str()))
                 ImGui::DockBuilderDockWindow(newScene->getName().c_str(), dockNode->ID);
         }
-        m_scenes[newScene->windowId] = newScene;
+        m_scenes[newScene->getWindowId()] = newScene;
     }
```
Done

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.cpp:
```diff
@@ -0,0 +1,103 @@
+		// No need to setup anything
+	}
+
+	void MaterialInspector::shutdown()
+	{
+		// No need to delete anything since the destructor of the framebuffer will handle it
+	}
```
I rather keep in case i need it

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.cpp:
```diff
@@ -0,0 +1,103 @@
+#include "exceptions/Exceptions.hpp"
+
+namespace nexo::editor {
+	void MaterialInspector::setup()
+	{
+		// No need to setup anything
+	}
```
Same as before

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.cpp:
```diff
@@ -0,0 +1,103 @@
+        }
+	}
+
+	void MaterialInspector::update()
+	{
+		// No need to update anything
+	}
```
same as before

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.cpp:
```diff
@@ -0,0 +1,103 @@
+			materialModified = false;
+			app.getSceneManager().deleteScene(previewParams.sceneId);
+		}
+		if (!m_framebuffer)
+			THROW_EXCEPTION(BackendRendererApiFatalFailure, "OPENGL", "Failedd to initialize framebuffer in Material Inspector window");
```
Done

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.hpp:
```diff
@@ -0,0 +1,36 @@
+
+namespace nexo::editor {
+
+	class MaterialInspector : public ADocumentWindow {
+		public:
+			using ADocumentWindow::ADocumentWindow;
+			void setup() override;
+			void show() override;
+        	void shutdown() override;
+         	void update() override;
+
+		private:
+			void renderMaterialInspector(int selectedEntity);
+
+			std::shared_ptr<renderer::Framebuffer> m_framebuffer = nullptr;
+			int m_ecsEntity = -1;
+			bool m_materialModified = true;
+	};
```
Should be done soon

iMeaNz commented on editor/src/WindowRegistry.hpp:
```diff
@@ -0,0 +1,55 @@
+			void update() const;
+			void render() const;
+
+		private:
+			std::unordered_map<std::type_index, std::shared_ptr<IDocumentWindow>> m_windows;
+
+			DockingRegistry m_dockingRegistry;
+	};
```
Not necessary

iMeaNz commented on editor/src/DocumentWindows/EntityProperties/RenderProperty.cpp:
```diff
@@ -23,26 +23,11 @@
-
-
-    void RenderProperty::createMaterialPopup(ecs::Entity entity)
+    void RenderProperty::createMaterialPopup(ecs::Entity entity) const
```
Should be fine

iMeaNz commented on editor/src/Editor.hpp:
```diff
@@ -64,23 +45,28 @@ namespace nexo::editor {
-            void registerWindow(const std::string& name, std::shared_ptr<IDocumentWindow> window);
-            void addLog(const LogMessage& message);
-            [[nodiscard]] const std::vector<LogMessage>& getLogs() const;
+            template<typename T>
+            requires std::derived_from<T, IDocumentWindow>
+            void registerWindow()
+            {
+            	auto window = std::make_shared<T>(m_windowRegistry);
+             	m_windowRegistry.registerWindow<T>(window);
+            }
+
```
Should be fine

iMeaNz commented on editor/src/DocumentWindows/MainScene.cpp:
```diff
@@ -96,11 +98,15 @@
 
-    void MainScene::shutdown()
-    {}
+    void MainScene::shutdown() const
+    {
+    	// Should probably check if it is necessary to delete the scene here ? (const for now)
+    }
```
I know but i rather keep for maybe future use

iMeaNz commented on editor/src/Editor.cpp:
```diff
@@ -228,30 +170,11 @@ namespace nexo::editor {
-    }
-
-    const std::vector<LogMessage> &Editor::getLogs() const
-    {
-        return m_logs;
+		setupEngine();
+		setupStyle();
+		m_windowRegistry.setup();
     }
```
Should be fine since we use exceptions

iMeaNz commented on editor/src/Editor.cpp:
```diff
@@ -265,31 +188,20 @@
     }
 
-    void Editor::buildDockspace() const
+    void Editor::buildDockspace()
     {
         const ImGuiViewport* viewport = ImGui::GetMainViewport();
         const ImGuiID dockspaceID = viewport->ID;
+        static bool dockingRegistryFilled = false;
 
```
Should be fine

iMeaNz commented:
@coderabbitai generate docstrings

iMeaNz commented:
@coderabbitai summary

Thyodas commented on editor/src/DocumentWindows/EntityProperties/AEntityProperty.hpp:
```diff
@@ -16,34 +16,30 @@
 
-            virtual void update() = 0;
-            virtual void showEnd() = 0;
+            virtual bool show(ecs::Entity entity) = 0;
```
Can you document with a docstring why it should return a bool, and what is it for?

iMeaNz replied:
Honestly no apparent reason, probably more of some sort of "coherence" with the way ImGui works.
Since it is a render function, it made sense initially to put a boolean here like ImGui does

iMeaNz replied:
I removed the boolean since it was not useful, see [718760d](https://github.com/NexoEngine/game-engine/pull/148/commits/718760d9d44b9193e857fd90b1b6b6d0fd4dc163)

Thyodas commented on editor/src/DocumentWindows/InspectorWindow.cpp:
```diff
@@ -39,77 +41,31 @@ extern ImGuiID g_materialInspectorDockID;
     {
-    	m_materialInspector->setup();
+		m_entityProperties[typeid(components::TransformComponent)] = std::make_shared<TransformProperty>(*this);
+		m_entityProperties[typeid(components::RenderComponent)] = std::make_shared<RenderProperty>(*this);
```
Maybe create a method to abstract this initialization
In this form:
```cpp
template <Component, Property>
    requires std::derives_from<Property, AEntityProperty>
void registerProperty()
{
    // Maybe handle if already existing?
    m_entityProperties[typeid(Component)] = std::make_shared<Property>(*this);
}
```

iMeaNz replied:
True

iMeaNz replied:
Implemented, see [20da8b5](https://github.com/NexoEngine/game-engine/pull/148/commits/20da8b5e6cfe11587699a2b380ab737038371708)

Thyodas commented on editor/src/WindowRegistry.hpp:
```diff
@@ -0,0 +1,120 @@
+			 * @throws WindowNotRegistered Thrown if no window of the specified type is registered.
+			 */
+			template<typename T>
+			std::shared_ptr<T> getWindow()
```
Maybe later we want to be able to unregister and delete a window (don't worry for now, maybe we can create an issue for it)
In the case we do that, a weak_ptr as return type would be useful to guarantee the validity of the pointer in case of deletion.

iMeaNz replied:
Yeah true too, since we don't have a use for deletion yet i did not think about it but this is something to be taken into account

iMeaNz replied:
The `getWindow` now returns a `weak_ptr`, see https://github.com/NexoEngine/game-engine/pull/148#issuecomment-2752387845

Thyodas reviewed (changes requested):
Overall LGTM, just minor suggestions

iMeaNz commented on editor/main.cpp:
```diff
@@ -28,15 +30,15 @@ int main(int argc, char **argv)
-        editor.registerWindow("Scene Tree", std::make_shared<nexo::editor::SceneTreeWindow>());
-        editor.registerWindow("Inspector", std::make_shared<nexo::editor::InspectorWindow>());
-        editor.registerWindow("Console", std::make_shared<nexo::editor::ConsoleWindow>(editor));
+        editor.registerWindow<nexo::editor::SceneViewManager>();
```
I guess multiple window types are not that common and should probably be treated as "exceptions" (like what we already do with the `SceneViewManager`).
Using types instead of strings prevents failure when for example a window name is determined by the user (like for the scenes for example), so that no matter what we can get the precise window when necessary.

Maybe if in the future it starts to be very common to have multiple windows of the same type at runtime it could be good because having multiple different managers could become a pain in the ass

iMeaNz replied:
As for the second point, I dont see the use case ? Like if you want to create a window that can be retrieved by another window just to be deleted right after, why not simply create it in the window itself and not register it ? Or did I miss something

Thyodas replied:
If a dev wants to open for example multiple Model windows, having a name or uuid is useful to specifically interact with one or another.
In blender, photoshop and so on I think it is possible to open multiple windows of the same type, for now we just don't have a use case in NEXO.

iMeaNz replied:
Implemented, see https://github.com/NexoEngine/game-engine/pull/148#issuecomment-2752387845

iMeaNz commented:
# Window Registry refactor 
The `WindowRegistry` now can hold multiple window of one type.
We store a
```cpp
std::unordered_map<std::type_index, std::vector<std::shared_ptr<IDocumentWindow>>>
```
The registration/getter function have also been modified to comply with this new setup
```cpp
template<typename T>
requires std::derived_from<T, IDocumentWindow>
void registerWindow(std::shared_ptr<T> window)
{
    auto &windowsOfType = m_windows[typeid(T)];
    if (std::ranges::any_of(windowsOfType, [&](const auto &existingWindow) {
            return existingWindow->getWindowName() == window->getWindowName();
        }))
    {
        THROW_EXCEPTION(WindowAlreadyRegistered, typeid(T), window->getWindowName());
    }
    windowsOfType.push_back(window);
}
```
We now also check if a window of the same type and the same name has already been registered (probably not good to throw an exception here but can be updated in the futur)
When we want to retrieve a specific window, we need to mention its name and its type.
```cpp
template<typename T>
requires std::derived_from<T, IDocumentWindow>
std::weak_ptr<T> getWindow(const std::string &windowName) const
{
    auto it = m_windows.find(typeid(T));
    if (it == m_windows.end())
	{
		LOG(NEXO_WARN, "Window of type {} not found", typeid(T).name());
			return {};
	}

    auto &windowsOfType = it->second;
    auto found = std::ranges::find_if(windowsOfType, [&windowName](const auto &w) {
        return w->getWindowName() == windowName;
    });

    if (found == windowsOfType.end())
	{
		LOG(NEXO_WARN, "Window of type {} with name {} not found", typeid(T).name(), windowName);
			return {};
	}

    return std::static_pointer_cast<T>(*found);
}
```
Also this function returns a weak_ptr to keep the ownership inside the `WindowRegistry` class
We can also retrieve all windows of a specific type like so 
```cpp
template<typename T>
requires std::derived_from<T, IDocumentWindow>
std::ranges::transform_view<
    std::ranges::ref_view<const std::vector<std::shared_ptr<IDocumentWindow>>>,
    std::shared_ptr<T>(*)(const std::shared_ptr<IDocumentWindow>&)
>
getWindows() const
{
    // Helper: non-capturing function for casting:
    std::shared_ptr<T>(*caster)(const std::shared_ptr<IDocumentWindow>&) = &castWindow<T>;

    auto it = m_windows.find(typeid(T));
    if (it == m_windows.end()) {
        static const std::vector<std::shared_ptr<IDocumentWindow>> empty;
        return std::ranges::transform_view(std::ranges::ref_view(empty), caster);
    }
    return std::ranges::transform_view(std::ranges::ref_view(it->second), caster);
}
```
This returns a transform_view on the vector of IDocument, allowing easy access when retrieving it (no need to cast it manually).

iMeaNz commented:
@coderabbitai review

iMeaNz commented:
# Editor Scene
With the `WindowRegistry` system, we don't need anymore the `SceneViewManager`.
So I refactored the `MainScene` class and renamed it `EditorScene` since it made more sense, this class now inherits from `ADocument` so it is handled like any other document window in the editor.
Not so much changed about it except for that

iMeaNz commented on editor/src/utils/String.cpp:
```diff
@@ -0,0 +1,27 @@
+//// String.cpp ///////////////////////////////////////////////////////////////
+//
+//  zzzzz       zzz  zzzzzzzzzzzzz    zzzz      zzzz       zzzzzz  zzzzz
+//  zzzzzzz     zzz  zzzz                    zzzz       zzzz           zzzz
+//  zzz   zzz   zzz  zzzzzzzzzzzzz         zzzz        zzzz             zzz
+//  zzz    zzz  zzz  z                  zzzz  zzzz      zzzz           zzzz
+//  zzz         zzz  zzzzzzzzzzzzz    zzzz       zzz      zzzzzzz  zzzzz
+//
+//  Author:      Mehdy MORVAN
+//  Date:        26/03/2025
+//  Description: Source file for the strings utils
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "String.hpp"
+
+namespace nexo::editor::utils {
+	std::string removeIconPrefix(const std::string &str)
+	{
+        size_t pos = str.find(" ");
+        if (pos != std::string::npos)
+        {
+            return str.substr(pos + 1);
+        }
+        return str;
+	}
```
Done

iMeaNz commented on editor/src/context/Selector.hpp:
```diff
@@ -44,6 +44,9 @@ namespace nexo::editor {
 			const std::string &getSelectedUuid() const;
 			void setSelectedEntity(std::string_view uuid, int entity);
 
+			void setSelectedScene(int scene);
+			int getSelectedScene() const;
```
Done

iMeaNz commented on editor/src/utils/TransparentStringHash.hpp:
```diff
@@ -0,0 +1,62 @@
+	     * @param s A null-terminated C-string to be hashed.
+	     * @return size_t The computed hash value for the input string.
+	     */
+	    size_t operator()(const char* s) const noexcept {
+	        return std::hash<std::string_view>{}(s);
+	    }
```
Done

iMeaNz commented on editor/src/DocumentWindows/EntityProperties/SpotLightProperty.hpp:
```diff
@@ -18,6 +18,19 @@
 	class SpotLightProperty : public AEntityProperty {
 		public:
-			static int show(ecs::Entity entity);
+			using AEntityProperty::AEntityProperty;
+
+			/**
+			 * @brief Renders the spotlight properties editor for the specified entity.
+			 *
+			 * Retrieves the spotlight component from the given entity and displays a user interface
+			 * for editing its properties, including color, direction, position, and cutoff parameters.
+			 * Adjusting the maximum distance automatically recalculates the linear and quadratic
+			 * attenuation factors, while the cutoff angles are presented in degrees for easier modification.
+			 *
+			 * @param entity The entity whose spotlight component will be edited.
+			 * @return true Always returns true, indicating that the properties editor was displayed successfully.
+			 */
+			void show(ecs::Entity entity) override;
```
Done

iMeaNz commented on editor/src/DocumentWindows/SceneTreeWindow.cpp:
```diff
@@ -204,8 +205,11 @@
-                    auto &viewManager = SceneViewManager::get();
-                    viewManager.addNewScene(sceneNameBuffer);
+                	//TOOD: create scene
+                    auto newScene = std::make_shared<EditorScene>(sceneNameBuffer, m_windowRegistry);
```
Will do later

iMeaNz commented on editor/src/DocumentWindows/SceneTreeWindow.cpp:
```diff
@@ -76,21 +80,17 @@
         return nodeOpen;
     }
 
     void SceneTreeWindow::sceneSelected(const SceneObject &obj) const
```
Fixed for now

iMeaNz commented on editor/main.cpp:
```diff
@@ -28,13 +29,16 @@ int main(int argc, char **argv)
-        editor.registerWindow("Scene Tree", std::make_shared<nexo::editor::SceneTreeWindow>());
-        editor.registerWindow("Inspector", std::make_shared<nexo::editor::InspectorWindow>());
-        editor.registerWindow("Console", std::make_shared<nexo::editor::ConsoleWindow>(editor));
+        //editor.registerWindow<nexo::editor::SceneViewManager>("Scene view manager");
+        editor.registerWindow<nexo::editor::EditorScene>("Default scene");
+        editor.registerWindow<nexo::editor::SceneTreeWindow>("Scene tree");
+        editor.registerWindow<nexo::editor::InspectorWindow>("Inspector");
+        editor.registerWindow<nexo::editor::ConsoleWindow>("Console");
+        editor.registerWindow<nexo::editor::MaterialInspector>("Material inspector");
+
```
Removed

iMeaNz commented on editor/src/DocumentWindows/EntityProperties/TransformProperty.hpp:
```diff
@@ -14,17 +14,24 @@
-            void update();
-
-            static int show(ecs::Entity entity);
+			/**
+			* @brief Displays and edits the transform properties of an entity using an ImGui interface.
+			*
+			* Retrieves the transform component (position, scale, and rotation quaternion) of the given entity,
+			* displaying the values in an ImGui table. The rotation is converted from a quaternion to Euler angles
+			* to allow intuitive editing; any changes in Euler angles are applied incrementally back to the quaternion,
+			* ensuring it remains normalized. The method adjusts UI styling to improve row spacing and returns a boolean
+			* indicating successful display.
+			*
+			* @param entity The entity whose transform properties are rendered.
+			*/
+            void show(ecs::Entity entity) override;
```
Done

iMeaNz commented on editor/src/DocumentWindows/EntityProperties/AEntityProperty.hpp:
```diff
@@ -16,34 +16,30 @@
 
-	class InterfaceEntityProperty {
-		public:
-			virtual ~InterfaceEntityProperty() = default;
-			virtual int show(ecs::Entity selectedEntity) = 0;
-	};
+namespace nexo::editor {
```
Done

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.hpp:
```diff
@@ -0,0 +1,57 @@
+
+namespace nexo::editor {
+
+	class MaterialInspector : public ADocumentWindow {
+		public:
+			using ADocumentWindow::ADocumentWindow;
+			void setup() override;
+
+			/**
+			 * @brief Displays the Material Inspector window.
+			 *
+			 * This method retrieves the selected entity and the inspector window, then renders the Material Inspector UI using ImGui.
+			 * It sets up appropriate window flags, performs a first-time docking setup when necessary, and delegates material
+			 * rendering to renderMaterialInspector() if the Material Inspector is visible.
+			 */
+			void show() override;
+        	void shutdown() override;
+         	void update() override;
```
Not necessary for now

iMeaNz commented on editor/src/DocumentWindows/MaterialInspector.hpp:
```diff
@@ -0,0 +1,57 @@
+			 */
+			void renderMaterialInspector(int selectedEntity);
+
+			std::shared_ptr<renderer::Framebuffer> m_framebuffer = nullptr;
+			int m_ecsEntity = -1;
+			bool m_materialModified = true;
```
There is already a check

iMeaNz commented on editor/src/DocumentWindows/ConsoleWindow.hpp:
```diff
@@ -49,16 +93,47 @@ namespace nexo::editor {
             };
 
-            const Editor& _editor;
+            std::vector<LogMessage> m_logs;
```
That might become a problem indeed, but probably better to fix it later.
Can you create an issue for that ?

iMeaNz commented on editor/src/DocumentWindows/ConsoleWindow.hpp:
```diff
@@ -49,16 +93,47 @@
-            const Editor& _editor;
+            std::vector<LogMessage> m_logs;
 
+            /**
+             * @brief Clears all log entries and display items.
+             *
+             * Removes all log messages from the internal storage and resets the list of display items.
+             */
             void clearLog();
+
+            /**
+             * @brief Appends a log message to the console's log collection.
+             *
+             * This method adds the provided log message to the internal container, ensuring it is available
+             * for display in the console window.
+             *
+             * @param message The log message to append.
+             */
+            void addLog(const LogMessage& message);
             void displayLog(loguru::Verbosity verbosity, const std::string &msg) const;
```
Will probably do it later

iMeaNz commented on editor/src/DocumentWindows/EditorScene.hpp:
```diff
@@ -0,0 +1,131 @@
+			*/
+            void deleteCamera(ecs::Entity cameraId);
+
+            void setDefault() { m_defaultScene = true; };
```
Not necessary

iMeaNz commented on editor/src/DocumentWindows/EntityProperties/RenderProperty.hpp:
```diff
@@ -20,17 +20,34 @@
-            static bool showMaterialInspector;
-            static components::Material *selectedMaterial;
-            static PopupManager popupManager;
+            /**
+             * @brief Displays a popup modal for creating a new material.
+             *
+             * Renders a two-column popup using ImGui where the left panel accepts material inputs (name and properties)
+             * via a material inspector, and the right panel shows a live preview of the material rendered in a framebuffer.
+             * If a preview scene hasn’t been generated yet, it is created for the specified dimensions. When the user confirms
+             * by clicking "OK", the material is linked to the entity's render component and the preview scene is cleaned up;
+             * clicking "Cancel" simply deletes the preview scene.
+             *
+             * @param entity The entity associated with the material being created.
+             */
+            void createMaterialPopup(ecs::Entity entity) const;
```
Its fine for now

iMeaNz commented on editor/src/ADocumentWindow.hpp:
```diff
@@ -16,11 +16,20 @@
 #include "Nexo.hpp"
+#include "WindowRegistry.hpp"
+
+#include <imgui_internal.h>
```
We track the updates

iMeaNz commented on editor/src/Editor.hpp:
```diff
@@ -14,73 +14,141 @@
+            requires std::derived_from<T, IDocumentWindow>
+            void registerWindow(const std::string &name)
+            {
+            	auto window = std::make_shared<T>(name, m_windowRegistry);
+             	m_windowRegistry.registerWindow<T>(window);
+            }
```
fine for now

iMeaNz commented on editor/src/DocumentWindows/EditorScene.cpp:
```diff
@@ -90,19 +87,23 @@
 
-    void MainScene::handleKeyEvents()
-    {}
+    void EditorScene::handleKeyEvents() const
+    {
+    	// Will be implemeneted later
+    }
```
Will do it later

iMeaNz commented on editor/src/DocumentWindows/EditorScene.cpp:
```diff
@@ -273,39 +265,55 @@
-            _viewPosition = ImGui::GetCursorScreenPos();
+            m_viewPosition = ImGui::GetCursorScreenPos();
+            firstDockSetup(m_windowName);
 
             m_focused = ImGui::IsWindowFocused();
             app.getSceneManager().getScene(m_sceneId).setActiveStatus(m_focused);
             if (m_focused)
             {
-                if (auto &viewManager = SceneViewManager::get();
-                    viewManager.getSelectedScene() != m_sceneId)
+                if (auto &selector = Selector::get();
+                    selector.getSelectedScene() != m_sceneId)
                 {
-                	auto &selector = Selector::get();
-                    viewManager.setSelectedScene(m_sceneId);
+                    selector.setSelectedScene(m_sceneId);
                     selector.unselectEntity();
                 }
             }
 
-            renderView();
-            renderGizmo();
+            if (m_activeCamera == -1)
+            {
+	            // No active camera, render the text at the center of the screen
+	            ImVec2 textSize = ImGui::CalcTextSize("No active camera");
+	            ImVec2 textPos = ImVec2((m_viewSize.x - textSize.x) / 2, (m_viewSize.y - textSize.y) / 2);
+
+	            ImGui::SetCursorScreenPos(textPos);
+	            ImGui::Text("No active camera");
+            }
+            else
+            {
+	            renderView();
+	            renderGizmo();
+            }
         }
         ImGui::End();
         ImGui::PopStyleVar();
     }
```
True, it is fixed

Thyodas commented on editor/src/DocumentWindows/InspectorWindow.hpp:
```diff
@@ -14,38 +14,164 @@
+			 *         or std::monostate if no data is available.
+			 */
+			template<typename T>
+			std::variant<std::monostate, components::Material*> getSubInspectorData() const
```
Why not use nullptr in case it doesn't exist, instead of encapsulating a pointer inside a variant?

iMeaNz replied:
Because this function is meant to be usable to get other types of data, idk let's say we have an animation component in the inspector where i can click to open an animation inspector, well the animation inspector will use this function to retrieve the animation component

Thyodas replied:
Yes but here you used a `components::Material*` as return type:
* why hardcoding this specific **Material** component as return type?
* if it is always a pointer, you can return a nullptr (here in the code it's hard coded as Material, it cannot change without modifying the declaration)

iMeaNz replied:
Oh I think I see what you mean, the variant was used initially because i store the data in a generic unordered_map container like so
```cpp
std::unordered_map<std::type_index, std::variant<std::monostate, components::Material *>> m_subInspectorData;
```
But I guess i can probably do something like :
```cpp
std::unordered_map<std::type_index, std::any> m_subInspectorData;
```
And then 
```cpp
template<typename WindowType, typename Data>
Data* getSubInspectorData() const
{
    auto it = m_subInspectorData.find(std::type_index(typeid(WindowType)));
    if (it != m_subInspectorData.end())
    {
        try {
            return std::any_cast<Data*>(it->second);
        }
        catch (const std::bad_any_cast& e) {
            return nullptr;
        }
    }
    return nullptr;
}

```

Like that it is not necessary for us to modify the signatures if we want to use other data

Thyodas replied:
Ok I see, yes we an do that later. Maybe using an interface instead of any? since it looks like it's always pointers we can use a void * or make components derive from an Interface.
@coderabbitai create an issue to improve this later please

iMeaNz replied:
Since its the ecs components that would be stored here, using an interface would probably mean to have to make derive every components from this base class. I guess this would not cause any real overhead but is it really necessary to change so much for such a small part of the code ?
void * is debatable, although it feels strange in cpp it can work

Thyodas commented on editor/src/DocumentWindows/EditorScene.cpp:
```diff
@@ -273,43 +191,54 @@ namespace nexo::editor {
     {
-        if (!m_opened)
+    	auto &selector = Selector::get();
+    	m_windowName = utils::removeIconPrefix(selector.getUiHandle(m_sceneUuid, m_windowName));
```
Why are we using a split to look for the name, what happens in case something is called "my windows", it removes "my"?
Why can't we seperate the icon from the real name in two seperate variables?

iMeaNz replied:
We can, but I was too lazy to refactor the way the scene tree handle the ui handles haha, but this is something I will update when adding new features to the scene tree

Thyodas reviewed (commented):
LGTM, just small questions
